### DISCUSSION TITLE
Building NORM for Android using gradle

### DISCUSSION THREAD

**Discussion by agrossmann09** [USNavalResearchLaboratory/norm#42] [Q&A]
I was able to build my .aar file using the gradle method supported, but when I attempt to import that .aar file I am getting the following error:

Null extracted folder for artifact: ResolvedArtifact(componentIdentifier=lib-release_-release.aar, variantName=null, artifactFile=/Users/Owner/AndroidStudioProjects/normApp/app/libs/lib-release_-release.aar, extractedFolder=null, dependencyType=ANDROID, isWrappedModule=false, buildMapping={__current_build__=/Users/Owner/AndroidStudioProjects/normApp}, mavenCoordinatesCache=com.android.build.gradle.internal.ide.dependencies.MavenCoordinatesCacheBuildService$Inject@6fae679f)

I could be importing the .aar file incorrectly or is there another library file I should be importing into my project?

**Comment by bebopagogo** [comment]
Have you been able to resolve this yet?

**Reply by agrossmann09** [reply]
Sorry for the late response. No I am still having issues building NORM for Android.

**Comment by bebopagogo** [comment]
What version of Android Studio, etc are you using?  Since I haven’t been focused on Android for   a couple of years, my Android setup may be a little out of date. I could try to update and see if there are issues, but I am able to build the Android native NORM libraries from the gradle project (after I added the missing files to the Github repo).  Since I mnainly use C++ frameworks for my development, I haven’t personally worked through the Android Java / jar package build with Android Studio.  We have older code that uses the older Android build tools to do that but I don’t think we have updated those build systems to use the Android Studio stuff (I think that may still be in the norm/makefiles/android-ndk directory but doesn’t work with current Android build tools as far as I know).  I can check with the developer here that was working that to see if he done something in that area.  That is for a separate “serverless” XMPP chat framework we created that uses NORM. That work was done in Java.

**Reply by agrossmann09** [reply]
I am working with Android Studio version 2020.3.1 Arctic Fox. I am able to build using Gradle but I can't locate any of the java classes. I have the .aar as I stated in the other thread. Please correct me if I am wrong since I haven't used CMake or Android Studio to generated linked libraries. Attached is a rough screen shot of my output. 

![Screen Shot 2021-11-03 at 12 48 55 PM](https://user-images.githubusercontent.com/64447380/140106045-b6d261d7-974d-41ff-9a58-cf20c9441346.png)

**Reply by agrossmann09** [reply]
And by native norm libraries do you mean the .aar?

**Comment by bebopagogo** [comment]
No – I mean the .so files that a Java .jar file would reference.  I _think_ the .aar is a sort of packaging of all of that (i.e. the .jar and any needed native .so libraries and maybe other artifacts to allow for easy incorporation of a package (like NORM) as dependency within another project).  I haven’t had need or experience with building the .aar since I don’t do Java Android development myself.   I did find some stuff on Stack Exchange or somewhere and thought I had provided a link to that to you in an earlier reply on this (i.e. how to add support to a gradle configuration to build the .aar … or at least hints towards that).  Probably some Android developer help forum might be useful for getting assistance on this.  I will be happy to incorporate something into the NORM repo here to support that if someone figures out what is needed.  I have not had time to pursue it myself.  I suspect it would be very straightforward for someone with the right knowledge or experience.  I will ask some of my team if they have any clues on this.
Cheers,
Brian

**Comment by bebopagogo** [comment]
Additionally, I think the JNI library would also need to be built in addition to the NORM native library.  I suspect a JAR file would be built by Android studio with addition of the JNI library build to the project.  I suspect that involves adding a stanza to the norm/android/lib/CMakelists.txt file that includes the source files located in norm/src/java/jni for building the NORM JNI library (for binding the NORM C/C++ API to the NORM Java classes that are located in norm/src/java/src.  Once you have the NORM native library, the norm-jni library  (and I assume one could statically link libnorm.a into the JNI shared library) and the JAR file that contains the NORM Java classes, you probably have all the ingredients needed the .aar package?  I'm speculating a little bit here but that may give you some clues about what needs to be done.

**Reply by agrossmann09** [reply]
Okay, correct me if I misunderstand, but I think the jni is already built into the .aar that Android Studio is creating. Under the .aar I see a jni folder. Within that jni folder I see the libmil_navy...so. And in the CMakeLists there is callout to the java/jni to include the .cpp files there. I think then I would just have to link the libnorm.a (is this packaged by a Makefile and if so would the linux makefile suffice for Android?), and build the JAR file? I think the only part I don't understand is where the JAR file would be coming from. The .aar packages some kind of classes.jar, but I don't think that is what I need.

**Comment by agrossmann09** [comment]
I also wanted to say in this discussion that I have a solution for the Android build if you wanted to take a look.

**Reply by bebopagogo** [reply]
Yes - I would be interested in that.

**Reply by agrossmann09** [reply]
How can I push my branch for you to review?

**Reply by weston-nrl** [reply]
Probably easiest to just submit a pull request

**Reply by agrossmann09** [reply]
I am trying to push to the branch that I created so that I can create a pull request but I am getting a authentication fails for https://github.com/USNavalResearchLaboratory/norm.git/

**Reply by weston-nrl** [reply]
You don't have permission to create a branch on the main repo.  Fork the repo into your own personal repository, do the work there, and commit a pull request from that.

**Reply by agrossmann09** [reply]
Gotcha. I have never pushed to a public repo before. I will submit that.

**Reply by agrossmann09** [reply]
Pull request submitted.

**Comment by bebopagogo** [comment]
I accepted your pull request and merged it in.  Thanks for the contribution!
